File: backend/app/schemas/spreadsheet.py
from __future__ import annotations
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class SpreadsheetCreate(BaseModel):
    title: str = Field(default="Untitled Spreadsheet", max_length=500)
    folder_id: uuid.UUID | None = None
    template_id: uuid.UUID | None = None


class SpreadsheetUpdate(BaseModel):
    title: str | None = Field(None, max_length=500)
    folder_id: uuid.UUID | None = None
    sheets_meta_json: dict | None = None


class SpreadsheetResponse(BaseModel):
    id: uuid.UUID
    title: str
    owner_id: uuid.UUID
    folder_id: uuid.UUID | None = None
    sheets_meta_json: dict | None = None
    row_count: int
    col_count: int
    is_template: bool
    template_category: str | None = None
    thumbnail_key: str | None = None
    is_trashed: bool
    last_edited_by: uuid.UUID | None = None
    last_edited_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend-app/src/types/spreadsheet.ts
export interface Spreadsheet {
  id: string;
  title: string;
  owner_id: string;
  folder_id: string | null;
  sheets_meta_json: SheetsMeta | null;
  row_count: number;
  col_count: number;
  is_template: boolean;
  template_category: string | null;
  thumbnail_key: string | null;
  is_trashed: boolean;
  last_edited_by: string | null;
  last_edited_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface SheetsMeta {
  sheets: SheetInfo[];
  cells?: Record<string, Record<string, CellData>>;
}

export interface SheetInfo {
  name: string;
  index: number;
  visible: boolean;
  color: string | null;
}

export interface CellData {
  value: string | number | boolean | null;
  formula?: string;
  format?: CellFormat;
}

export interface CellFormat {
  bold?: boolean;
  italic?: boolean;
  underline?: boolean;
  fontFamily?: string;
  fontSize?: number;
  textColor?: string;
  backgroundColor?: string;
  textAlign?: "left" | "center" | "right";
  verticalAlign?: "top" | "middle" | "bottom";
  numberFormat?: string;
  wrapText?: boolean;
}
